### DISCUSSION TITLE
appium can't switch to dual app's webview

### DISCUSSION THREAD

**Discussion by gbcbooks** [appium/appium#17749] [Q&A]
i have a wechat cloned 
originalPackageName : com.tencet.mm
dual PackageName: dkplugin.gly.fyv (it is radmon,i can't control ) 

start the dkplugin.gly.fyv , and start mini APP

from the appium driver.contexts, it's output is 
`>>> driver.contexts
['NATIVE_APP', 'WEBVIEW_com.tencent.mm:appbrand0', 'WEBVIEW_com.tencent.mm:appbrand2']`
when driver.switch_to.context('WEBVIEW_com.tencent.mm:appbrand0')
get error 
`>>> driver.switch_to.context("WEBVIEW_com.tencent.mm:appbrand0")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/gbcbooks/.local/lib/python3.10/site-packages/appium/webdriver/switch_to.py", line 43, in context
    self._driver.execute(MobileCommand.SWITCH_TO_CONTEXT, {'name': context_name})
  File "/home/gbcbooks/.local/lib/python3.10/site-packages/selenium/webdriver/remote/webdriver.py", line 429, in execute
    self.error_handler.check_response(response)
  File "/home/gbcbooks/.local/lib/python3.10/site-packages/appium/webdriver/errorhandler.py", line 30, in check_response
    raise wde
  File "/home/gbcbooks/.local/lib/python3.10/site-packages/appium/webdriver/errorhandler.py", line 26, in check_response
    super().check_response(response)
  File "/home/gbcbooks/.local/lib/python3.10/site-packages/selenium/webdriver/remote/errorhandler.py", line 243, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: An unknown server-side error occurred while processing the command. Original error: A new session could not be created. Details: session not created: please close 'com.tencent.mm' and try again
Stacktrace:
UnknownError: An unknown server-side error occurred while processing the command. Original error: A new session could not be created. Details: session not created: please close 'com.tencent.mm' and try again
    at getResponseForW3CError (/usr/local/lib/node_modules/appium/node_modules/appium-base-driver/lib/protocol/errors.js:804:9)
    at asyncHandler (/usr/local/lib/node_modules/appium/node_modules/appium-base-driver/lib/protocol/protocol.js:380:37)`
from the error code , it mention '**com.tencent.mm**',  not 'dkplugin.gly.fyv', i am wondering , can't appium handle  dual cloned APP ?

**Comment by jlipps** [comment]
I'm not sure. But the client side log is not very useful. The entire Appium server log would be more helpful.

**Reply by gbcbooks** [reply]
yes , here is the appium server log 
https://gist.github.com/gbcbooks/42eda0b70394f6a28ccbdc388e24c4ac

**Reply by KazuCocoa** [reply]
Potentially deprecated of https://github.com/appium/appium/issues/16514

Could you try out with latest UIA2 driver with Appium 2 once?

**Reply by gbcbooks** [reply]
ok, let me try

**Reply by gbcbooks** [reply]
hello , i try appium2 , appium2 server can find dual package's webview content now

but i am facing a new problem 

if open two app ,  and bot of them get into mini app, driver.contexts will show both of mini app from this two app, just like the following 
```
>>> driver.window_handles
['CDwindow-BB816FB404DCBAF7929F95CB8B1ECC8C', 'CDwindow-264A1E15239E172D4E3973CB11FF478C']
>>> driver.contexts
['NATIVE_APP', 'WEBVIEW_com.tencent.mm:appbrand3', 'WEBVIEW_com.tencent.mm:appbrand0', 'WEBVIEW_com.tencent.mm:appbrand2', 'WEBVIEW_com.tencent.mm:appbrand0']
>>> 
```

it shows two ''WEBVIEW_com.tencent.mm:appbrand0', when i driver.swith_to.context(''WEBVIEW_com.tencent.mm:appbrand0'), i don't know which context i have switch to.

**Reply by KazuCocoa** [reply]
https://github.com/appium/appium-uiautomator2-driver/#mobile-getcontexts provides you more WebView information. I guess this will help you to find a better one.
If one webview does not have an exact page, `appium:ensureWebviewsHavePages` as false also might help to filter them out
https://github.com/appium/appium-uiautomator2-driver/#web-context
